api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java

```diff
@@ -387,18 +406,181 @@ public boolean supportsPropertyName(String gpName) {
 		}
 	}
 
+	private Patient persistImportedMpiPatient(MpiPatient mpiPatient) {
+		// TODO: Fix bug MPI-11 Updating of Locally generated ID in MPI after MPI import not occurring
+		String openMrsIdTypeUuid = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_UUID);
+
+		PatientIdentifierType openMrsIdType = patientService.getPatientIdentifierTypeByUuid(openMrsIdTypeUuid);
+		Patient patient;
+		if (mpiPatient.getPatientIdentifier(openMrsIdType) == null) {
+			PatientIdentifier localId = validateOrGenerateIdentifier(null, null);
+			mpiPatient.addIdentifier(localId);
+		}
+		patient = patientService.savePatient(mpiPatient);
+		return patient;
+	}
+
+	private IdentifierSourceService getIssAndUpdateIdSource() {
+		IdentifierSourceService iss = Context.getService(IdentifierSourceService.class);
+		if (idSource == null) {
+			String idSourceId = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID);
+			if (StringUtils.isBlank(idSourceId)) {
+				throw new APIException("Please set the id of the identifier source to use to generate patient identifiers");
+			}
+			try {
+				idSource = iss.getIdentifierSource(Integer.valueOf(idSourceId));
+				if (idSource == null) {
+					throw new APIException("cannot find identifier source with id:" + idSourceId);
+				}
+			}
+			catch (NumberFormatException e) {
+				throw new APIException("Identifier source id should be a number");
+			}
+		}
+
+		return iss;
+	}
+
+	private Location getIdentifierLocation(Location identifierLocation) {
+		if (identifierLocation == null) {
+			identifierLocation = locationService.getDefaultLocation();
+			if (identifierLocation == null) {
+				throw new APIException("Failed to resolve location to associate to patient identifiers");
+			}
+		}
+
+		// see if there is a primary identifier location further up the chain, use that instead if there is
+		Location identifierAssignmentLocationAssociatedWith = getIdentifierAssignmentLocationAssociatedWith(identifierLocation);
+		if (identifierAssignmentLocationAssociatedWith != null) {
+			identifierLocation = identifierAssignmentLocationAssociatedWith;
+		}
+
+		return identifierLocation;
+	}
+
+	private PatientIdentifier validateOrGenerateIdentifier(String identifierString, Location identifierLocation) {
+		IdentifierSourceService iss = getIssAndUpdateIdSource();
+		identifierLocation = getIdentifierLocation(identifierLocation);
+
+		// generate identifier if necessary, otherwise validate
+		if (StringUtils.isBlank(identifierString)) {
+			identifierString = iss.generateIdentifier(idSource, null);
+		} else {
+			PatientIdentifierValidator.validateIdentifier(identifierString, idSource.getIdentifierType());
+		}
+
+		PatientIdentifier pId = new PatientIdentifier(identifierString, idSource.getIdentifierType(),
+				identifierLocation);
+		pId.setPreferred(true);
+
+		return pId;
+	}
+
+	private String getGlobalProperty(String propertyName) {
```

Comment:
I think you defined a method like this in another earlier class, why not define a public one in a utility class so you can reuse it in all these places.

Reply:
Removed because nothing was using it anyways